proposals/xxxx-mesh-nodes.md

```diff
@@ -0,0 +1,302 @@
+is the replacement of the input payload with a node input record, and the use
+of shader attributes, including a new launch type: `[NodeLaunch("mesh")]`.
+
+This is an HLSL-focused summary of the graphics nodes feature proposed in the
+[work graphs spec](https://github.com/microsoft/DirectX-Specs/blob/master/d3d/WorkGraphs.md#graphics-nodes).
+For now only mesh node support is considered as an experimental addition.
+
+## Motivation
+
+Work graphs are a great way to generate compute work on the GPU, and there is a
+lot of interest in expanding this to scheduling graphics work as a replacement
+for execute indirect.
+Mesh shader is a natural place to start, as it's already a compute shader, and
+has far fewer details to work out than supporting legacy graphics pipelines.
+
+## Proposed solution
+
+Add a new node launch type `"mesh"`, indicating a mesh shader leaf graphics node.
+The new mesh node is based on a combination of a broadcast launch node and a
+mesh shader. See more details about the runtime context in the work graphs spec
+[here](https://github.com/microsoft/DirectX-Specs/blob/master/d3d/WorkGraphs.md#graphics-nodes).
+
+Use the node input record as the mesh shader input payload, instead of an
+explicit `payload` argument.
+Like a broadcast node input record, this must have an `SV_DispatchGrid` field
+indicating the grid size when `[NodeMaxDispatchGrid(x,y,z)]` is used.
+
+The same system value inputs supported for broadcast launch nodes and mesh
+shaders are supported in the same way here:
+`SV_DispatchThreadID`, `SV_GroupThreadID`, `SV_GroupIndex`, and `SV_GroupID`.
+
+Node outputs are not allowed - the mesh node must be a leaf in the work graph.
+
+The new `mesh` node supports outputs inherited from the mesh shaders:
+`vertices`, `primitives`, and `indices`.  As with mesh shaders, the
+`SetMeshOutputCounts` intrinsic must be used before writing to these outputs.
+As with mesh shaders, the `vertices` and `primitives` outputs are combined to
+provide the inputs the pixel shader requires.
+Each of these require a user defined structure using semantics on fields, where
+fields of HLSL basic types will be used to generate signature elements in the
+same way they are for mesh shader signature handling.
+
+Amplification shaders are not supported in this context, since work graph node
+shaders provide the function of generating the mesh node dispatches on the GPU.
+
+Mesh nodes can also be used as one or more entrypoints in a work graph
+(standalone since they can't output to other nodes),
+in which case the input record (payload) is supplied from DispatchGraph()
+at the API just like any other work graph entrypoint input.
+
+A complete mesh node is defined at the API by first creating a
+[generic program](https://github.com/microsoft/DirectX-Specs/blob/master/d3d/WorkGraphs.md#d3d12_generic_program_desc)
+including the mesh node function name, a pixel shader, and other optional state,
```

Comment:
What are these optional states?

Reply:
You need to see the Work Graphs spec for complete details here.  This was just an attempt to briefly summarize the context into which this shader function fits.  There are other subobject types for defining parts of the graphics pipeline state which could be added to the program as well.  See subobject types listed [here](https://github.com/microsoft/DirectX-Specs/blob/master/d3d/WorkGraphs.md#d3d12_state_subobject_type) for some relevant graphics-related state that might be used in addition to the shaders.

Reply:
One example is a pixel shader, which can be defined via ps_6_*.  Other examples are rasterizer state, blend state, depth stencil state, long list.  In the future we should add support for defining all those subobjects in DXIL so they don't have to be defined at the API.  For now, one has to define some things in the API since not everything can be expressed in HLSL.  As an example there is support for defining raytracing subobjects in HLSL already. The runtime extracts these and turns them in to DDI subobject descriptions that are equivalent to API ones, so we wouldn't even need drivers to change.

Reply:
It feels like more than just mesh here. Is it possible to call this something like graphics node instead of mesh node?

Reply:
This was basically a "dispatchmesh" launch node, which was shortened to "mesh" because it was cleaner and there wasn't a good reason not to.  DispatchMesh launches a graphics pipeline of a certain kind, which does require associated state.  The mesh node shader defined here is really just the mesh node shader portion, requiring the rest of the state to set up a graphics pipeline (or "program") to run.

The term "graphics node" would imply more generality which is proposed in the work graphs spec (see various links), including the ability to launch the full legacy graphics pipeline, starting with some vertex shader-like node instead of a mesh shader-like node.  Even those would need different NodeLaunch types to indicate the type of the pipeline, such as "draw" or "drawindexed", then "mesh" is still needed to distinguish the node type being defined here from the others.

This proposal focuses only on defining the mesh shader part of the graphics nodes proposal.  Hence, "mesh launch node" or "mesh node" for short.